### DISCUSSION TITLE
How to initialize part of a model from a checkpoint in JAX/Flax?

### DISCUSSION THREAD

**Discussion by marcvanzee** [google/flax#923] [Q&A] [answered]

**Comment by marcvanzee** [comment, marked as answer]
Suppose you have a model called `BigModel`. You want to load part of the weights of this model from a checkpoint, and you want to initialize the rest of the weights randomly. For instance, you may want to use BERT-encoded text as input. 

You can do this by initializing the entire model with random weights, and then replacing the subtree of parameters corresponding to the parameters from your checkpoint. In Flax, (immutable) parameters are represented as a nested [FrozenDict](https://github.com/google/flax/blob/master/flax/core/frozen_dict.py) (just a dict that ensures immutability, so that it's harder to shoot yourself in the foot with JAX that assumes pure functions). So before you can modify the parameters, you have to make sure you "unfreeze" them, and "freeze" them again after modification.

Pseudo-code:

```
from flax.core import freeze, unfreeze

bert_params = load_bert_params()  # These are the params you want to load.
params = BigModel(...).init(...)['params']
params = unfreeze(params)
params['bert'] = bert_params
params = freeze(params)
```

Suppose you would like to do a "soft update" with smoothing coefficient `t`, so for every param `x` it does something like `x = t * cur_x + (1-t) * new_x`. In that case you can simply use `jax.tree_multimap` to apply an update fn to all the leaves of the tree:

```
A = {'a':jnp.zeros((5,)), 'b':jnp.zeros((5,))}
B = {'a':jnp.ones((5,)), 'b':jnp.ones((5,))}
t=0.1
jax.tree_multimap(lambda x,y: x*(1-t) + y*t, A, B)
#{'a': DeviceArray([0.1, 0.1, 0.1, 0.1, 0.1], dtype=float32),
# 'b': DeviceArray([0.1, 0.1, 0.1, 0.1, 0.1], dtype=float32)}
```

So you can simply do `params = jax.tree_multimap(lambda x, y: x*(1-t) + y*t, params, new_params)`.

**Comment by clee1994** [comment]
Hey, 

I have a somewhat similar question.

How would you do something like data-driven initialization? Do you have a suggestion on how to load parts of a model then call init and initialize the rest of the model? 

My use case is that I want to quantize a model with pre-trained weights and then train quantization parameters (step size for example - https://arxiv.org/abs/1902.08153). So I want to load pre-trained weights and then run the initializer on new parameters in the network, such as quantization step size (here it would also be important that data is passed through the network, so I can capture statistics). 

Any ideas on how to do that? Any help is appreciated!

Best,
Clemens

**Reply by marcvanzee** [reply]
Internally, `init` just calls `apply` and sets the variable collection `params` to mutable. So if you want to mutate some data-dependent variables based on existing weights, you can use `apply` and set the variable collection you want to update to `mutable`. If there are any mutated variables, `apply` will return a pair `(output, mutated_var)` and you can use that to update the variable dict.

This is maybe clearer with an example:

```python
from flax import linen as nn
import jax.numpy as jnp
from jax import random
from flax.linen import initializers
from flax.core.frozen_dict import freeze, unfreeze

class Foo(nn.Module):
  @nn.compact
  def __call__(self, x):
    y = self.param('pretrained', initializers.zeros, (2, 3))
    z = self.variable('data-driven', 'z', jnp.ones, (2, 3))
    z.value = y + z.value
    return x + y + z.value

variables = Foo().init(random.PRNGKey(0), jnp.zeros((2, 3)))
print(f'Initial variables: {variables}')
```

```
Initial variables: FrozenDict({
    params: {
        pretrained: DeviceArray([[0., 0., 0.],
                     [0., 0., 0.]], dtype=float32),
    },
    data-driven: {
        z: DeviceArray([[1., 1., 1.],
                     [1., 1., 1.]], dtype=float32),
    },
})
```

```python
pretrained_weights = {'pretrained': jnp.zeros((2, 3))+10}

variables = unfreeze(variables)
variables["params"].update(pretrained_weights)
variables = freeze(variables)

print(f'Variables with pretrained weights:', variables)
```

```
Variables with pretrained weights: FrozenDict({
    data-driven: {
        z: DeviceArray([[1., 1., 1.],
                     [1., 1., 1.]], dtype=float32),
    },
    params: {
        pretrained: DeviceArray([[10., 10., 10.],
                     [10., 10., 10.]], dtype=float32),
    },
})
```

```python
_, mutated_vars = Foo().apply(variables, jnp.zeros((2, 3)),mutable=["data-driven"])

variables = unfreeze(variables)
variables.update(mutated_vars)
variables = freeze(variables)

print(f'Variables after initializing:', variables)
```

```
Variables after initializing: FrozenDict({
    data-driven: {
        z: DeviceArray([[11., 11., 11.],
                     [11., 11., 11.]], dtype=float32),
    },
    params: {
        pretrained: DeviceArray([[10., 10., 10.],
                     [10., 10., 10.]], dtype=float32),
    },
})
```

Let me know if that answers your question!

**Reply by clee1994** [reply]
Thank you @marcvanzee that helped a lot, keeping quantization parameters separate as variables was a very good idea and solved my problem. :)

**Comment by cgarciae** [comment]
This issue is related to #1619. I think doing Transfer Learning is currently a bit painful for beginners.
